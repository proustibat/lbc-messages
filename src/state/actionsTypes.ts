export enum messagesActionTypes {
  INIT = '@@INIT',
  REQUEST_MESSAGES = 'REQUEST_MESSAGES',
  REQUEST_MESSAGES_SUCCESS = 'REQUEST_MESSAGES_SUCCESS',
  REQUEST_MESSAGES_FAILURE = 'REQUEST_MESSAGES_FAILURE',
}
