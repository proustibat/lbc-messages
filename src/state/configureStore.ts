import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rootReducer from './reducers';
import { RootStateOrAny } from 'react-redux';

const middleware = [thunk];

const configureStore = (preloadedState?: RootStateOrAny) => {
  return createStore(
    rootReducer,
    preloadedState,
    composeWithDevTools(applyMiddleware(...middleware)),
  );
};

export default configureStore;
