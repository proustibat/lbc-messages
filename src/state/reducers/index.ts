import { combineReducers } from 'redux';
import messages, {MessagesReducerType} from './messagesReducer';

export type RootReducersType = {
  messages: MessagesReducerType
};
export default combineReducers({
  messages,
});
