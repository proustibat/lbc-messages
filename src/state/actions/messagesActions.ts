import { messagesActionTypes as Actions } from '../actionsTypes';
export const API_URL_GET_MESSAGES = '/01ce1c8d';
export const API_URL_POST_MESSAGE = '/b32b976b';

export const requestMessages = () => ({
  type: Actions.REQUEST_MESSAGES,
  payload: {
    request: {
      url: API_URL_GET_MESSAGES,
      method: 'GET',
    },
  },
});

export const sendMessage = (data: { title: string; message: string }) => ({
  type: Actions.SEND_MESSAGE,
  payload: {
    request: {
      url: API_URL_POST_MESSAGE,
      method: 'POST',
      data,
    },
  },
});
