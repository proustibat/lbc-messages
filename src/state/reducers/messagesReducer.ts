import { AxiosResponse } from 'axios';
import { messagesActionTypes as Actions } from '../actionsTypes';
import { MessageProps } from '../../components/Message';

export type MessagesReducerType = {
  items: MessageProps[];
  loading: boolean;
};

export const initialState: MessagesReducerType = {
  items: [],
  loading: false,
};

const messagesReducer = (
  state = initialState,
  action: {
    type: Actions;
    payload?: { data?: unknown; request?: AxiosResponse };
  },
) => {
  switch (action.type) {
    case Actions.REQUEST_MESSAGES:
      return {
        ...state,
        loading: true,
      };
    case Actions.REQUEST_MESSAGES_FAILURE:
      return {
        ...state,
        loading: false,
      };
    case Actions.REQUEST_MESSAGES_SUCCESS:
      return {
        ...state,
        items: (action?.payload?.data as { messages: MessageProps[] })
          ?.messages,
        loading: false,
      };
    case Actions.SEND_MESSAGE_SUCCESS:
      return {
        ...state,
        items: [
          ...state.items,
          action.payload?.data
        ],
        loading: false,
      };
    default:
      return state;
  }
};

export default messagesReducer;
