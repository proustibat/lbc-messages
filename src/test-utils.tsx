import React, { ReactNode } from 'react';
import configureStore, { MockStore } from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { RootReducersType } from './state/reducers';

export const getMockedStore = (
  initialState: RootReducersType = { messages: { items: [], loading: false } },
) => {
  const createMockStore = configureStore([thunk]);
  const mockStore: MockStore = createMockStore(initialState);
  return mockStore;
};

export const withProvider = (WrappedComponent: React.ComponentType<any>) => (
  initialState: RootReducersType,
) => (props: any) => {
  return (
    <Provider store={getMockedStore(initialState)}>
      <WrappedComponent {...props} />
    </Provider>
  );
};

export const withRouter = (children: any) => {
  return (
    <MemoryRouter>
      <Route>{children}</Route>
    </MemoryRouter>
  );
};
