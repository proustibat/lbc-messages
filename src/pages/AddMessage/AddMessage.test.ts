import {
  act,
  fireEvent,
  render,
  cleanup,
} from '@testing-library/react';
import { AddMessage } from './index';
import { messages } from '../../fixtures/messages';
import { RootReducersType } from '../../state/reducers';
import { fillForm, withProvider, withRouter } from "../../test-utils";
import * as messagesActions from '../../state/actions/messagesActions';

const stateLoadingMessages: RootReducersType = {
  messages: { items: [], loading: true },
};

const stateWithMessages: RootReducersType = {
  messages: { items: messages, loading: false },
};

const stateWithoutMessages: RootReducersType = {
  messages: { items: [], loading: false },
};

describe('AddMessage view', () => {
  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });
  afterAll(() => {
    // jest.restoreAllMocks()
  });
  describe.each([
    ['the store has messages', stateWithMessages],
    ['the store has not any messages', stateWithoutMessages],
    ['the store is loading messages', stateLoadingMessages],
  ])('given %s', (_, state) => {
    it('should render the view correctly by default', () => {
      // Given
      const ProvidedAddMessage = withProvider(AddMessage)(state);

      // When
      const { container } = render(withRouter(ProvidedAddMessage));

      // Then
      expect(container).toMatchSnapshot();
    });
  });

  it('should handle the form submit', async () => {
    // Given
    const sendMessageSpy = jest.spyOn(messagesActions, 'sendMessage');
    const ProvidedAddMessage = withProvider(AddMessage)(stateWithMessages);
    const { getByTestId } = render(withRouter(ProvidedAddMessage));

    // When
    await fillForm(getByTestId, { title: 'title', message: 'message' }, true);

    // Then
    expect(sendMessageSpy).toHaveBeenCalledTimes(1);
    sendMessageSpy.mockRestore();
  });
});
