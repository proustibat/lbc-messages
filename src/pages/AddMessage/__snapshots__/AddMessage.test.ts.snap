// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMessage view given the store has messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h4
      class="MuiTypography-root title_padding-32px 0px 16px 0px MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
    >
      add-new-message
    </h4>
    <div
      class="MuiBox-root MuiBox-root-1 formContainer_marginTop-16"
    >
      <div
        class="MuiBox-root MuiBox-root-2 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
      >
        <form
          autocomplete="off"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
            data-testid="input-title"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="title"
              id="title-label"
            >
              title
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="title"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-5"
                >
                  <span>
                    title
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
            data-testid="input-message"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="message"
              id="message-label"
            >
              message
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                id="message"
                rows="4"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-5"
                >
                  <span>
                    message
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
            data-testid="submit-button"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              send
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root rightIcon_marginLeft-16"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                />
              </svg>
            </span>
          </button>
        </form>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-7 homeButton_position-fixed homeButton_left-0 homeButton_bottom-16 homeButton_width-100%"
    >
      <button
        aria-label="home"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`AddMessage view given the store has not any messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h4
      class="MuiTypography-root title_padding-32px 0px 16px 0px MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
    >
      add-new-message
    </h4>
    <div
      class="MuiBox-root MuiBox-root-8 formContainer_marginTop-16"
    >
      <div
        class="MuiBox-root MuiBox-root-9 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
      >
        <form
          autocomplete="off"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
            data-testid="input-title"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="title"
              id="title-label"
            >
              title
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="title"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-10 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-12"
                >
                  <span>
                    title
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
            data-testid="input-message"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="message"
              id="message-label"
            >
              message
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                id="message"
                rows="4"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-10 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-12"
                >
                  <span>
                    message
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
            data-testid="submit-button"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              send
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root rightIcon_marginLeft-16"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                />
              </svg>
            </span>
          </button>
        </form>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-14 homeButton_position-fixed homeButton_left-0 homeButton_bottom-16 homeButton_width-100%"
    >
      <button
        aria-label="home"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`AddMessage view given the store is loading messages should render the view correctly by default 1`] = `
<div>
  <div
    class="intro_backgroundColor-#fff intro_padding-32px 0px 32px"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
      >
        home-title
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        home-intro
      </p>
    </div>
  </div>
  <div
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
    role="progressbar"
  >
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
    />
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
    />
  </div>
  <div
    class="MuiContainer-root root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h4
      class="MuiTypography-root title_padding-32px 0px 16px 0px MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-alignCenter MuiTypography-displayBlock"
    >
      add-new-message
    </h4>
    <div
      class="MuiBox-root MuiBox-root-15 formContainer_marginTop-16"
    >
      ...
    </div>
    <div
      class="MuiBox-root MuiBox-root-16 homeButton_position-fixed homeButton_left-0 homeButton_bottom-16 homeButton_width-100%"
    >
      <button
        aria-label="home"
        class="MuiButtonBase-root MuiFab-root root_backgroundColor-#3f51b5 root_color-#fff"
        data-testid="floating-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
